src/pipx/commands/install.py

```diff
@@ -57,7 +56,7 @@ def install(
         venv = Venv(venv_dir, python=python, verbose=verbose)
         venv.check_upgrade_shared_libs(pip_args=pip_args, verbose=verbose)
         if exists:
-            if not reinstall and force and python_flag_was_passed:
+            if not reinstall and force:
```

Comment:
Would this now not always print the warning when `--force` is passed, even when `--python` is not passed by design? I understood that is the exact thing the issue was supposed to adress?

Reply:
Sorry, I misunderstood the issue. How do you suggest I go about it ? This warning should only pop up if `--force` and `--python` are used together right ?

Reply:
Yes, exactly.

My first hunch would be to change the logic for `python_flag_was_passed` to
```python
python_flag_was_passed = python != DEFAULT_PYTHON
```
This is only an approximation though. If the user explicitly provided the default python via the `--python` flag, this would not print the warning.

The other option, as described in the issue, is to not set `DEFAULT_PYTHON` as the default argument to the `--python` flag. That way, we can actually check, if the flag was set or not.

If it does not make the code a lot more complex, I'd lean towards option 2, as it's more accurate.

Reply:
Yes that's the one.

You have to check all usages of `args.python` and see if it's replaced with `DEFAULT_PYTHON` somewhere down the line. Just setting it to `None` and running the test suite should already give you a decent idea if it might be a problem, as most of the test suite does not use the `--python` flag.

Reply:
Tests fails at this point, after the change, as python is `None`
```python3
    m.update(python.encode())
```
Do you think we should have a check for `python` being `None` here ?

Reply:
How about setting the default value to an empty string?

Reply:
That causes `InterpreterResolutionError` in almost all tests.

```python3
DEBUG    pipx.main:main.py:207 Failed to resolve interpreter:
Traceback (most recent call last):
  File "/Users/swaraj/personal/pipx/src/pipx/main.py", line 204, in run_pipx_command
    interpreter = find_python_interpreter(args.python, fetch_missing_python=fetch_missing_python)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/swaraj/personal/pipx/src/pipx/interpreter.py", line 72, in find_python_interpreter
    raise InterpreterResolutionError(source="PATH", version=python_version)
pipx.interpreter.InterpreterResolutionError: No executable for the provided Python version '' found in PATH. Please make
sure the provided version is on your PATH or the file path is valid. The
provided version looks like a version for Python Launcher, but `py` was not
found on PATH.
```

Reply:
OK, maybe empty string is not a good option. Thanks for the trying.

Reply:
Then I think need to do something like `python = python or DEFAULT_PYTHON` [1] in the corresponding places.

https://github.com/pypa/pipx/blob/ac70dc0d2221bc6a78e39bef86f43b5c73ef03d7/src/pipx/commands/install.py#L38

Reply:
> Then I think need to do something like `python = python or DEFAULT_PYTHON` [1] in the corresponding places.
> 
> https://github.com/pypa/pipx/blob/ac70dc0d2221bc6a78e39bef86f43b5c73ef03d7/src/pipx/commands/install.py#L38

Yes, this.

Reply:
> [1] in the corresponding places.

Does this mean everywhere we are using args.python or just fixing it `run_pipx_command` will do (see the snippet below) ? Sorry if I misunderstood the comment.
```python3
interpreter = find_python_interpreter(args.python or DEFAULT_PYTHON, fetch_missing_python=fetch_missing_python)
```

Reply:
> Does this mean everywhere we are using args.python or just fixing it `run_pipx_command` will do (see the snippet below) ? Sorry if I misunderstood the comment.
> 
> ```python
> interpreter = find_python_interpreter(args.python or DEFAULT_PYTHON, fetch_missing_python=fetch_missing_python)
> ```

Doing it here will be fine I think. You'll have to store to a variable whether or not `args.python` was set before this line though, so you can pass it to where it's needed for the check.

Reply:
Oh OK, so I can do something like this then ?
```python3
        ...
        python_flag_was_passed = args.python or DEFAULT_PYTHON
        try:
            interpreter = find_python_interpreter(args.python, python_flag_was_passed, fetch_missing_python=fetch_missing_python)
        ...
```
And then change `find_python_interpreter` accordingly ?
```python3
def find_python_interpreter(python_version: str, python_flag_was_passed: bool = True, fetch_missing_python: bool = False) -> str:
    if not python_flag_was_passed:
        return DEFAULT_PYTHON
    ...
```

Reply:
I think the point where this line is written, `args.python` will always be set (refer below). But if we make the above fix, it will let us set the default value to an empty string as it will pick up `DEFAULT_PYTHON` at this point.
```python3
if "python" in args and args.python is not None:
        fetch_missing_python = args.fetch_missing_python
        python_flag_was_passed = bool(args.python)
        try:
            interpreter = find_python_interpreter(args.python, python_flag_was_passed, fetch_missing_python=fetch_missing_python)
            args.python = interpreter
        except InterpreterResolutionError as e:
            logger.debug("Failed to resolve interpreter:", exc_info=True)
            print(
                pipx_wrap(
                    f"{hazard} {e}",
                    subsequent_indent=" " * 4,
                )
            )
            return EXIT_CODE_SPECIFIED_PYTHON_EXECUTABLE_NOT_FOUND
```

Reply:
We also will need to change `find_python_interpreter` 's signature a bit. Something like this
```python3
def find_python_interpreter(python_version: str, python_flag_was_passed: bool = True, fetch_missing_python: bool = False) -> str:
    if not python_flag_was_passed:
        return DEFAULT_PYTHON
    ...
```

Reply:
Let me know if this fix works

Reply:
Why not just change it to?

```python
    if "python" in args:
        fetch_missing_python = args.fetch_missing_python
        python_flag_passed = bool(args.python)
        try:
            interpreter = find_python_interpreter(args.python or DEFAULT_PYTHON, fetch_missing_python=fetch_missing_python)
            args.python = interpreter
        except:
            ...
```

Reply:
no, but we need to create it here as the information whether the flag was set or not will be lost afterwards, due to
```python
        try:
            interpreter = find_python_interpreter(args.python or DEFAULT_PYTHON, fetch_missing_python=fetch_missing_python)
            args.python = interpreter
```

Reply:
Ah okay, makes sense. Made the changes